apps/goapp/app.go

```diff
@@ -0,0 +1,211 @@
+package goapp
+
+import (
+	"bytes"
+	"io/fs"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"unicode"
+
+	"github.com/gorilla/mux"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+	"github.com/mattermost/mattermost-plugin-apps/utils/httputils"
+)
+
+type App struct {
+	Manifest apps.Manifest
+
+	log    utils.Logger
+	Mode   apps.DeployType
+	Router *mux.Router
+
+	command       *BindableMulti
+	postMenu      []Bindable
+	channelHeader []Bindable
+}
+
+type AppOption func(app *App) error
+
+func MakeAppOrPanic(m apps.Manifest, opts ...AppOption) *App {
+	app, err := MakeApp(m, opts...)
+	if err != nil {
+		panic(err)
+	}
+	return app
+}
+
+func MakeApp(m apps.Manifest, opts ...AppOption) (*App, error) {
+	// Default the app's permissions
+	if len(m.RequestedPermissions) == 0 {
+		m.RequestedPermissions = []apps.Permission{
+			apps.PermissionActAsBot,
+		}
+	}
+
+	app := &App{
+		Manifest: m,
+		Router:   mux.NewRouter(),
+	}
+
+	// Run the options.
+	for _, opt := range opts {
+		err := opt(app)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Set up the auto-served HTTP routes.
+	app.Router.Path("/ping").HandlerFunc(httputils.DoHandleJSONData([]byte("{}")))
+	app.Router.Path("/manifest.json").HandlerFunc(httputils.DoHandleJSON(&app.Manifest)).Methods("GET")
+	app.HandleCall("/bindings", app.getBindings)
+
+	app.Router.NotFoundHandler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		app.log.Debugf("App request: not found: %q", req.URL.String())
+		http.NotFound(w, req)
+	})
+
+	return app, nil
+}
+
+func WithLog(log utils.Logger) AppOption {
+	return func(app *App) error {
+		app.log = log
+		return nil
+	}
+}
+
+func WithStatic(staticFS fs.FS) AppOption {
+	return func(app *App) error {
+		app.Router.PathPrefix("/static/").Handler(http.FileServer(http.FS(staticFS)))
+		return nil
+	}
+}
+
+func WithCommand(subcommands ...Bindable) AppOption {
```

Comment:
I'm thinking it would be beneficial to have this sort of library/driver implemented in another language as well, such as Javascript or Python. Maybe we should recommend this for the Apps currently being created in other languages. The shorthand notation of other languages could make this very useful.

Reply:
WithCommand _registers_ a possible command with the runtime, the selection of what to include in the /bindings request is still on the per-context basis, and can be further customized by the app by implementing its own `Bindable`s. Its really just a callback, with a "standard" handler for the Require flags